### PR TITLE
VerrazzanoHelidon CRD

### PR DESCRIPTION
**Pull request by arjav-desai**
This is change contains CRD and workload related integration along with empty/no-op implementation of controller.

### CODE REVIEW

File: application-operator/config/crd/patches/cainjection_in_verrazzanohelidonworkloads.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2021 Oracle and/or its affiliates.
```
Comment (mgianatagh) [inline application-operator/config/crd/patches/cainjection_in_verrazzanohelidonworkloads.yaml:1]: I do not think this file gets used, so we should not check it in.

@kminder I feel there is more than one file in this PR that was generated when a new API was added but is never used.  I will note them all.  My feeling is they should not be added to the repo.

File: application-operator/config/crd/patches/webhook_in_verrazzanohelidonworkloads.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2021 Oracle and/or its affiliates.
```
Comment (mgianatagh) [inline application-operator/config/crd/patches/webhook_in_verrazzanohelidonworkloads.yaml:1]: I do not think this file gets used, so we should not check it in.

File: application-operator/config/rbac/verrazzanohelidonworkload_editor_role.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2021 Oracle and/or its affiliates.
```
Comment (mgianatagh) [inline application-operator/config/rbac/verrazzanohelidonworkload_editor_role.yaml:1]: I do not think this file gets used, so we should not check it in.

File: application-operator/config/rbac/verrazzanohelidonworkload_viewer_role.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2021 Oracle and/or its affiliates.
```
Comment (mgianatagh) [inline application-operator/config/rbac/verrazzanohelidonworkload_viewer_role.yaml:1]: I do not think this file gets used, so we should not check it in.

File: application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2021 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+apiVersion: oam.verrazzano.io/v1alpha1
+kind: VerrazzanoHelidonWorkload
+metadata:
+  name: verrazzanohelidonworkload-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (mgianatagh) [inline application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml:10]: Should this sample be more functional?

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2021 Oracle and/or its affiliates.
```
Comment (mgianatagh) [inline platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml:1]: I do not think this file gets used, so we should not check it in.

File: application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+package v1alpha1
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// VerrazzanoHelidonWorkloadSpec wraps meta/ObjectMeta & apps/DeploymentSpec.
+type VerrazzanoHelidonWorkloadSpec struct {
+	// +kubebuilder:validation:EmbeddedResource
```
Comment (fredtibbitts) [inline application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go:13]: This should be removed I think. You don't have API version and kind.

File: application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	// +kubebuilder:pruning:PreserveUnknownFields
+	ObjectMeta metav1.ObjectMeta `json:"objectMeta"`
+	DeploymentSpec appsv1.DeploymentSpec `json:"deploymentSpec"`
```
Comment (fredtibbitts) [inline application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go:16]: gofmt on these 2 lines?

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml (hunk 6 above)
Comment (fredtibbitts) [reply platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml:1]: This file is used, it gets loaded at install time and if it's missing the application operator fails to start.

**Review by fredtibbitts — CHANGES_REQUESTED** [review]

**Comment by fredtibbitts** [comment]
Looks like the branch build failed due to gofmt errors. You should run `make check` to do gofmt, vet, lint checks. Unfortunately `check` isn't a dependency of any of the other build targets so you have to remember to run it explicitly. We should fix the Makefile.

File: application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go (hunk 7 above)
Comment (kminder) [reply application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go:13]: Assuming that CR creation has been tested and works why do you think this annotation needs to be removed?

File: application-operator/config/crd/bases/oam.verrazzano.io_verrazzanocoherenceworkloads.yaml [hunk 9]
```diff
@@ -37,8 +37,9 @@ spec:
           metadata:
             type: object
           spec:
-            description: VerrazzanoCoherenceWorkloadSpec wraps a Coherence resource. The Coherence object must include apiVersion,
-              kind, and spec fields. It may include a metadata field.
+            description: VerrazzanoCoherenceWorkloadSpec wraps a Coherence resource.
+              The Coherence object must include apiVersion, kind, and spec fields.
```
Comment (kminder) [inline application-operator/config/crd/bases/oam.verrazzano.io_verrazzanocoherenceworkloads.yaml:41]: Is the apiVersion, kind part still true?

File: application-operator/config/crd/patches/cainjection_in_verrazzanohelidonworkloads.yaml (hunk 1 above)
Comment (kminder) [reply application-operator/config/crd/patches/cainjection_in_verrazzanohelidonworkloads.yaml:1]: +1.  We do have to clean up the others but no sense in making the problem worse.

File: application-operator/config/crd/patches/webhook_in_verrazzanohelidonworkloads.yaml (hunk 2 above)
Comment (kminder) [reply application-operator/config/crd/patches/webhook_in_verrazzanohelidonworkloads.yaml:1]: +1.  We do have to clean up the others but no sense in making the problem worse.

File: application-operator/config/rbac/verrazzanohelidonworkload_editor_role.yaml (hunk 3 above)
Comment (kminder) [reply application-operator/config/rbac/verrazzanohelidonworkload_editor_role.yaml:1]: +1.  We do have to clean up the others but no sense in making the problem worse.

File: application-operator/config/rbac/verrazzanohelidonworkload_viewer_role.yaml (hunk 4 above)
Comment (kminder) [reply application-operator/config/rbac/verrazzanohelidonworkload_viewer_role.yaml:1]: +1.  We do have to clean up the others but no sense in making the problem worse.

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	vzapi "github.com/verrazzano/verrazzano/application-operator/apis/oam/v1alpha1"
+	// +kubebuilder:scaffold:imports
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller_test.go:21]: What is this annotation for?

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+var cfg *rest.Config
+var k8sClient client.Client
+var testEnv *envtest.Environment
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"Controller Suite",
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller_test.go:35]: Should this be helidon workload specific?

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2021 Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+package helidonworkload
+
+import (
+	"path/filepath"
+	"testing"
+
+	. "github.com/onsi/ginkgo"
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller_test.go:10]: Eliminate dot imports.

File: application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go (hunk 7 above)
Comment (fredtibbitts) [reply application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go:13]: That annotation tells k8s to validate that the embedded resource contains an api version and kind, but we're removing those, so the annotation should be removed.

File: application-operator/main.go [hunk 13]
```diff
@@ -25,6 +19,14 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	vzapi "github.com/verrazzano/verrazzano/application-operator/apis/oam/v1alpha1"
```
Comment (kminder) [inline application-operator/main.go:23]: These seem out of alphabetical order now.

File: platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml (hunk 6 above)
Comment (kminder) [reply platform-operator/helm_config/charts/verrazzano-application-operator/crds/oam.verrazzano.io_workloaddefinition_vzhelidon.yaml:1]: +1 to Fred's comment.

File: application-operator/config/crd/bases/oam.verrazzano.io_verrazzanocoherenceworkloads.yaml (hunk 9 above)
Comment (fredtibbitts) [reply application-operator/config/crd/bases/oam.verrazzano.io_verrazzanocoherenceworkloads.yaml:41]: This change has nothing to do with Helidon, looks like just a formatting change. It should probably be removed from the PR although I don't think it hurts to leave it here. When I submit a PR to remove api version and kind from VerrazzanoCoherenceWorkload I'll be updating the description to remove those.

File: application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml (hunk 5 above)
Comment (arjav-desai) [reply application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml:10]: Adding example based on CRD in the next commit.

File: application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go (hunk 7 above)
Comment (arjav-desai) [reply application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go:13]: +1 to what Fred said so removing it in next commit.

File: application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go (hunk 8 above)
Comment (arjav-desai) [reply application-operator/apis/oam/v1alpha1/verrazzanohelidonworkload_types.go:16]: done with next commit.

File: application-operator/config/crd/bases/oam.verrazzano.io_verrazzanocoherenceworkloads.yaml (hunk 9 above)
Comment (arjav-desai) [reply application-operator/config/crd/bases/oam.verrazzano.io_verrazzanocoherenceworkloads.yaml:41]: Thanks Fred, yes, its due to `make check` run. I will leave it as is for now.

File: application-operator/main.go (hunk 13 above)
Comment (arjav-desai) [reply application-operator/main.go]: Fixed the import order.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+	//TODO: find the resource object to record the event to, default is the parent appConfig.
+
+	//unwrap the apps/DeploymentSpec and meta/ObjectMeta
+
+	log.Info("Successfully created Verrazzano Helidon workload")
+	return ctrl.Result{}, nil
+}
+
+// locateParentAppConfig locate the parent application configuration object
+/*
```
Comment (kminder) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:64]: Should this commented out code be deleted?

File: application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml [hunk 15]
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2021 Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+apiVersion: oam.verrazzano.io/v1alpha1
+kind: VerrazzanoHelidonWorkload
+metadata:
+  name: verrazzanohelidonworkload-sample
+spec:
+  deploymentMeta:
```
Comment (kminder) [inline application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml:9]: The use of deploymentMeta and deploymentSpec don't seem to align with the equivalents in other workloads, right?

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go (hunk 10 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller_test.go]: I replaced the auto-generated test file so this is gone now.

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go (hunk 12 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller_test.go]: I replaced the auto-generated test file so this is gone now.

File: application-operator/controllers/helidonworkload/helidonworkload_controller_test.go (hunk 11 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller_test.go]: I replaced the auto-generated test file so this is gone now.

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go (hunk 14 above)
Comment (arjav-desai) [reply application-operator/controllers/helidonworkload/helidonworkload_controller.go]: I had started work on next PR so had commented it out. But removed it to make this PR clean.

File: application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml (hunk 15 above)
Comment (arjav-desai) [reply application-operator/config/samples/oam_v1alpha1_verrazzanohelidonworkload.yaml]: I replaced this with deploymentTemplate as we discussed.

**Comment by fredtibbitts** [comment]
I think you also need to add your CRD to the platform operator helm config, otherwise the application operator install will fail.

**Review by fredtibbitts — APPROVED** [review]

**Pull request merged**